File: c-scalar/src/main/java/com/scalar/graphqldgs/datafetchers/ShowsDatafetcher.java
package com.scalar.graphqldgs.datafetchers;

import com.netflix.graphql.dgs.DgsComponent;
import com.netflix.graphql.dgs.DgsQuery;
import com.netflix.graphql.dgs.InputArgument;
import com.scala.graphqldgs.types.Show;
import com.scalar.graphqldgs.services.ShowsService;

import java.util.List;
import java.util.stream.Collectors;

@DgsComponent
public class ShowsDatafetcher {
    private final ShowsService showsService;
    
    public ShowsDatafetcher(ShowsService showsService) {
        this.showsService = showsService;
    }
    
    /**
     * This datafetcher resolves the shows field on Query.
     * It uses an @InputArgument to get the titleFilter from the Query if one is defined.
     */
    @DgsQuery
    public List<Show> shows(@InputArgument("titleFilter") String titleFilter) {
        if (titleFilter == null) {
            return showsService.shows();
        }
        
        return showsService.shows().stream().filter(s -> s.getTitle().contains(titleFilter)).collect(Collectors.toList());
    }
}

File: x-kotlin/src/main/kotlin/com/kotlin/xkotlin/datafetchers/ShowsDataFetcher.kt
package com.kotlin.xkotlin.datafetchers

import com.example.demo.generated.types.Show
import com.kotlin.xkotlin.services.ShowsService
import com.netflix.graphql.dgs.DgsComponent
import com.netflix.graphql.dgs.DgsQuery
import com.netflix.graphql.dgs.InputArgument
import kotlinx.coroutines.coroutineScope

@DgsComponent
class ShowsDataFetcher(private val showsService: ShowsService) {
    @DgsQuery
    suspend fun shows(@InputArgument titleFilter: String?): List<Show> = coroutineScope {
        if (titleFilter != null) {
            showsService.shows().filter { it.title.contains(titleFilter) }
        } else {
            showsService.shows()
        }
    }
}